Plutarch/Api/V1/AssocMap.hs

```diff
@@ -56,3 +63,37 @@ instance
       x' <- Plutus.fromData x
       y' <- Plutus.fromData y
       Just (x', y')
+
+-- | Look up the given key in a 'PMap'.
+lookup :: (PIsData k, PIsData v) => Term (s :: S) (k :--> PMap k v :--> PMaybe v)
+lookup = phoistAcyclic $
+  plam $ \key ->
+    lookupDataWith
+      # (phoistAcyclic $ plam $ \pair -> pcon $ PJust $ pfromData $ psndBuiltin # pair)
+      # pdata key
+
+-- | Look up the given key data in a 'PMap'.
+lookupData :: (PIsData k, PIsData v) => Term (s :: S) (PAsData k :--> PMap k v :--> PMaybe (PAsData v))
+lookupData = lookupDataWith # (phoistAcyclic $ plam $ \pair -> pcon $ PJust $ psndBuiltin # pair)
+
+-- | Look up the given key data in a 'PMap', applying the given function to the found key-value pair.
+lookupDataWith ::
```

Comment:
I think it might be more efficient to have an explicit impl here, something of this sort perhaps
```hs
plookup :: (PIsData k, POrd k) => Term s (PAsData k :--> PMap k v :--> PMaybe (PAsData v))
plookup = phoistAcyclic $
  plam $ \kdat m ->
    plet (pfromData kdat) $ \k ->
      precList
        ( \self x xs -> unTermCont $ do
            k' <- pletC $ pfstBuiltin # x
            pure $
              pif
                (pfromData k' #< k)
                (self # xs)
                $ pif
                  (kdat #== k')
                  (pcon . PJust $ psndBuiltin # x)
                  $ pcon PNothing
        )
        (const $ pcon PNothing)
        # pto m
```

Reply:
That does improve performance:

      Actual:   
      valueOf {"exBudgetCPU":12794940,"exBudgetMemory":28740,"scriptSizeBytes":222}
      Expected: 
      valueOf {"exBudgetCPU":12106088,"exBudgetMemory":29008,"scriptSizeBytes":229}

However PlutusTx's [lookup](https://playground.plutus.iohkdev.io/doc/haddock/plutus-tx/html/PlutusTx-AssocMap.html#v:lookup) does not require `Ord k` and I think we should mirror the semantics of PlutusTx API.

I've used your `precList` idea to get

      valueOf {"exBudgetCPU":10676984,"exBudgetMemory":24208,"scriptSizeBytes":200}

Reply:
I think we're gonna have to decide whether we go for the invariant assumptions labeled by #344 anyway. In which case, we probably are going to have `POrd` in everything. I do imagine the optimization of finding a key through comparison along the way will be pretty bad for smaller maps and values though.

Also in general, I don't think mirroring PlutusTx's API is a good idea FWIW. There's a lot of inefficiency warts that don't work well for UPLC. But maybe we can afford to not have peak efficiency everywhere, that's upto others to decide.

Reply:
There is no reason to mirror what PlutusTx does.